src/impl_x86__base_implementation.inl

```diff
@@ -570,6 +573,25 @@ X86Microarchitecture GetX86Microarchitecture(const X86Info* info) {
         return X86_UNKNOWN;
     }
   }
+  if (IsVendorByX86Info(info, CPU_FEATURES_VENDOR_CENTAUR_HAULS) || 
```

Comment:
are you sure we should merge both ?

for me Centaur Hauls should be separated with the switch for 
samuel CPU aka `CENTAUR_SAMUEL`

ref: https://en.wikichip.org/wiki/centaur/microarchitectures/samuel

Reply:
before its aquisition by VIA 
> In September 1999, Centaur was purchased from IDT by VIA Technologies, a Taiwanese company.
ref: https://en.wikipedia.org/wiki/Centaur_Technology#History

Which have continued the family with the `samuel2` etc...
https://en.wikichip.org/wiki/via_technologies/microarchitectures/samuel_2

@gchatelet  should we also add a `is_via` `VIA_SAMUEL2` etc. ?

Concerning zhaoxin
> is a fabless semiconductor company, created in 2013 as a joint venture between VIA Technologies and the Shanghai Municipal Government
ref: https://en.wikipedia.org/wiki/Zhaoxin

so if we need a merge, it is with `VIA` IMHO...

Reply:
I fully agree with @Mizux that we should separate.

#### Note: 
Samuel, Samuel2 etc.., same cpuid vendor, "CentaurHauls"
refs:
http://instlatx64.atw.hu/
[Cpuid Dump Via Cyrix III (WinChip 5A, C5A, Samuel)l](http://users.atw.hu/instlatx64/CentaurHauls/CentaurHauls0000663_C5A_Samuel_CPUID.txt)
[Cpuid Dump Via Cyrix III (WinChip 5B, C5B, Samuel2)](http://users.atw.hu/instlatx64/CentaurHauls/CentaurHauls0000673_C5B_Samuel2_CPUID.txt)
[Cpuid Dump QuadCore Zhaoxin KaiXian KX-6640MA](http://users.atw.hu/instlatx64/CentaurHauls/CentaurHauls00307B0_6640MA_CPUID.txt)

Reply:
At present, the Vendor ID of CentaurHauls belongs to Zhaoxin, whose CPU products will use both CentaurHauls and Shanghai. Before the establishment of Zhaoxin, VIA CPU also used the Vendor ID of CentaurHauls, but VIA CPU was not mass produced, and there was no VIA CPU in use on the market. Moreover, we were not very clear about the details of VIA CPU, so we could not support VIA CPU here. We don't think it makes much sense to distinguish VIA CPUs here.

Reply:
I would prefer to separate `CPU_FEATURES_VENDOR_CENTAUR_HAULS` and `CPU_FEATURES_VENDOR_SHANGHAI` since, this provides flexibility, for example some "shangai" uarch's cpuid is the same as "centaur" uarch's cpuid. By splitting, we can determine the appropriate microarchitecture, in case that vendor's cpuid strings are merged
then you need to either split this condition or process the cpuid vendor string, I think it is a very rare case, but such a case should be considered.

Reply:
Yes, I will separate CPU_FEATURES_VENDOR_CENTAUR_HAULS and CPU_FEATURES_VENDOR_SHANGHAI.

Reply:
At the moment the same cpuid check for `CPU_FEATURES_VENDOR_CENTAUR_HAULS` and  `CPU_FEATURES_VENDOR_SHANGHAI`, I meant we should completely separate centaur and shanghai including uarch.
e.g we know that for uarch [ZHANGJIAN](http://users.atw.hu/instlatx64/CentaurHauls/CentaurHauls00006FE_CNR_Isaiah_CPUID.png) cpuid vendor is centaur hauls, thus we have to add this check only to `IsVendorByX86Info(info, CPU_FEATURES_VENDOR_CENTAUR_HAULS)`